proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,513 @@
+5. If the property wrapper defines a `projectedValue`, a local computed property representing  `projectedValue` will be synthesized by the compiler and named per the original parameter name prefixed with a dollar sign (`$`). The same accessor rules for `wrappedValue` apply to `projectedValue`.
+6. When passing an argument to a property wrapper parameter, the compiler will wrap the argument in the appropriate `init(wrappedValue:)` call.
+
+#### Transformation Example:
+
+```swift
+@propertyWrapper
+struct Percentage {
+
+  init(wrappedValue: Int) { ... }
+    
+    
+  var wrappedValue: Int {
+    get { ... }
+    set { ... }
+  }
+  
+}
+
+func reportProgress(
+  @Percentage of progress: Int
+) { ... }
+
+
+reportProgress(of: 50)
+```
+
+In the above code, the `reportProgress(of:)` function and its caller are equivalent to:
+
+```swift
+func reportProgress(of _progress: Percentage) {
+
+  var progress: Int {
+    get { _progress.wrappedValue }
+    // The setter accessor is not synthesized
+    // because the setter of Percentage's  
+    // 'wrappedValue' is mutating.
+  }
+
+
+  ...
+  
+}
+
+
+reportProgress(of: Percentage(wrappedValue: 50))
+```
+
+#### Restrictions on Property Wrapper Function Parameters
+
+##### `@autoclosure`
+
+Function parameters with a property wrapper custom attribute cannot have an `@autoclosure` type. `@autoclosure` is unnecessary for the wrapped value itself because the wrapped value argument at the call-site will always be wrapped in a call to `init(wrappedValue:)`, which already can support `@autoclosure` arguments. Using `@autoclosure` for the backing wrapper type would be misleading, because the type spelled out for the parameter would not be the true autoclosure type. Consider the `reportProgress` example from above with an `@autoclosure` parameter:
+
+```swift
+func reportProgress(
+  @Percentage progress: @autoclosure () -> Int
+) { ... }
+
+
+reportProgress(of: 50)
+```
+The above code would be transformed to:
+```swift
+func reportProgress(
+  progress _progress: @autoclosure () -> Percentage<Int>
+) { ... }
+
+
+reportProgress(of: Percentage(wrappedValue: 50))
+```
+The changing type of the `@autoclosure` is incredibly misleading, as it's not obvious that `@autoclosure` applies to the backing wrapper rather than the wrapped value. Therefore, using `@autoclosure` for a property wrapper function parameter will be a compiler error.
+
+
+### Property Wrappers on Closure Parameters
+
+Since a property wrapper custom attribute is applied directly to the declaration that will be wrapped, application of a property wrapper type is only available within a closure expression. That is, the signature of a function that contains a closure cannot include the property wrapper attibute. Instead the application of the attrbute will be up to the caller of the function, which supplies the closure argument.
```

Comment:
```suggestion
Since a property wrapper custom attribute is applied directly to the declaration that will be wrapped, application of a property wrapper type is only available within a closure expression. That is, the signature of a function that contains a closure cannot include the property wrapper attribute. Instead the application of the attribute will be up to the caller of the function, which supplies the closure argument.
```